### DISCUSSION TITLE
Intellisense - Dim Inactive Regions: Where does vscode-cpptools get the values for variables used in "#IF" e.g. "GTK_CHECK_VERSION"?

### DISCUSSION THREAD

**Discussion by ToxicDragon** [microsoft/vscode-cpptools#8523] [Q&A]
Hi there,
I have the problem that vscode-cpptools "dim inactive regions" option dims the region in the statement
`#if GTK_CHECK_VERSION(4,0,0)`
and additionally Intellisense also does not work in that region.

When changing the statement to
`#if GTK_CHECK_VERSION(3,0,0)`
the respective region is not dimmed anymore and also IntelliSense works again.

However when running the application, the following code
`cout << gtk_get_major_version() << "." << gtk_get_minor_version() << "." << gtk_get_micro_version() << endl;`
returns `4.5.0`
Therefore, I know for a fact that at compile time level my application uses GTK 4.5.0.

So where does the vscode-cpptools know the GTK version from and therefore is able to judge `GTK_CHECK_VERSION(4,0,0)`?

I also looked into my project's `.vscode` folder into the `c_cpp_properties.json` file and added the GTK4 library install dir `"/usr/local/include/gtk-4.0"` as follows:
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${default}",
                "/usr/local/include/gtk-4.0",
                "/usr/include/gtk-3.0",
                "/usr/include/glib-2.0",
                "/usr/lib/x86_64-linux-gnu/glib-2.0/include",
                "/usr/include/pango-1.0"
            ],
            "defines": [],
            "cStandard": "c17",
            "intelliSenseMode": "linux-clang-x64"
        }
    ],
    "version": 4
}
```
But this did not have the desired effect; still the respective code block is dimmed and Intellisense does not work.

"So why not turn off the feature "Dim inactive regions"?", I thought. Well, this makes the code undimmed (yay) BUT IntelliSense STILL does not work (nay), which is a major problem for me.

Can you help me out or is it a bug that IntelliSense does not work EVEN if the "dim inactive regions" option is disabled?

**Comment by sean-mcmanus** [comment]
Look at the definition GTK_CHECK_VERSION -- I'm guessing you may need to add some "defines".

**Reply by ToxicDragon** [reply]
GTK_CHECK_VERSION is a function of the GTK toolkit, see https://docs.gtk.org/gtk4/func.check_version.html
So no #defines needed. In other chats about this topic one clue was pointed out that the problem may reside in compile_commands.json of the project I am compiling (LibreOffice). I am currently investigating further.

**Reply by sean-mcmanus** [reply]
GTK_CHECK_VERSION appears to be a macro -- https://docs.gtk.org/gtk4/func.CHECK_VERSION.html . Can you see what it's defined as? I'm guessing you may need to set GTK_MAJOR_VERSION to 4, and/or look into how that is being set to 4 when you compile.

**Comment by ToxicDragon** [comment]
However, as far as I can see, IntelliSense not working when the option "Dim Inactive Regions" is disabled, might be considered a bug.

**Comment by Colengms** [comment]
Hi @ToxicDragon .  Can you explain what you mean by:

> the respective region is not dimmed anymore and also *IntelliSense works again*.

When I set`"C_Cpp.dimInactiveRegions": false`, I do not get IntelliSense in a disabled block.  For example:

![image](https://user-images.githubusercontent.com/49173979/145884645-10447f53-0d6c-4784-8cec-eceba01d1aa9.png)

Note the lack of a squiggle in the first block.  IntelliSense is based on parsing/compiling the code. If IntelliSense sees a disabled block, it does not parse its contents.

Can you provide a simple, self-contained example that demonstrates the issue?
